lib/core/src/display/navigation/camera_manager.rs

```diff
@@ -0,0 +1,197 @@
+use crate::prelude::*;
+
+use super::mouse_manager::State;
+use super::mouse_manager::MouseManager;
+use super::mouse_manager::MouseClickEvent;
+use super::mouse_manager::MouseWheelEvent;
+use super::mouse_manager::MousePositionEvent;
+use super::mouse_manager::MouseButton;
+use crate::system::web::Result;
+use crate::display::rendering::DOMContainer;
+
+use nalgebra::Vector2;
+use std::rc::Rc;
+use std::cell::RefCell;
+use nalgebra::zero;
+
+// ====================
+// === ZoomingEvent ===
+// ====================
+
+/// A struct holding zooming event information.
+pub struct ZoomingEvent {
+    pub focus  : Vector2<f32>,
+    pub amount : f32
+}
+
+impl ZoomingEvent {
+    fn from_mouse_wheel(event:MouseWheelEvent, data:&Rc<RefCell<CameraManagerData>>) -> Self {
```

Comment:
I don't particulary like passing `Rc<RefCell<...>>` around. It is error prone. Do you have for example any guarantee that the function which calls this function did not borrow it mutably in the past? Much better way is to enclose it in another structure and provide public API which is safer than that (getters and setters which does not allow you to enter mutable borrow and do another action inside). For example, if you define a struct which has a setter, then the mutable borrow will be done internally on the beginning of the setter and will be freed as soon as the setter stops executing, so the public API will always be safe. Let's just agree that we never pass `Rc<RefCell<...>>` explicitly until there is a real reason for that. Can we change it to something safe here?